### PR TITLE
[otbn,dv] RIG Code Cleanup

### PR DESCRIPTION
**Pull request by ctopal**
Code now makes all the shape checks before returning the instruction
from mnemonic.

Signed-off-by: Canberk Topal <ctopal@lowrisc.org>

### CODE REVIEW

File: hw/ip/otbn/dv/rig/rig/snippet_gen.py [hunk 1]
```diff
@@ -98,6 +100,156 @@ def pick_weight(self,
         '''
         return 1.0
 
+    def _check_insn_shape(self, insn: Optional[Insn], mnemonic: str) -> bool:
```
Comment (rswarbrick) [inline hw/ip/otbn/dv/rig/rig/snippet_gen.py:103]: This looks like it should work, but the giant dictionary is a bit messy. How about something this shape?
```python3
    @staticmethod
    def _check_bnxor_shape(ops: List[int]) -> bool:
        return (len(ops) == 6 and
                isinstance(ops[0].op_type, RegOperandType) and
                ops[0].op_type.reg_type == 'wdr' and
                ops[0].op_type.is_dest() and
                isinstance(ops[1].op_type, RegOperandType) and
                ops[1].op_type.reg_type == 'wdr' and
                not ops[1].op_type.is_dest() and
                isinstance(ops[2].op_type, RegOperandType) and
                ops[2].op_type.reg_type == 'wdr' and
                not ops[2].op_type.is_dest() and
                isinstance(ops[4].op_type, ImmOperandType))

    def _check_insn_shape(self, insn: Insn, mnemonic: str) -> bool:
        shape_checkers = {
            'bn.xor': SnippetGen._good_bnxor_shape,
            # TODO: Add the others here!
        }
        checker = shape_checkers.get(mnemonic)
        return checker is None or checker(insn.operands)

    def _get_named_insn(self, insns_file: InsnsFile, mnemonic: str) -> Insn:
        '''Get an instruction from insns_file by mnemonic

        This is used for specialized snippets that need to generate a specific
        instruction and wraps the error handling for when someone has removed
        the instruction from the file.

        '''
        insn = insns_file.mnemonic_to_insn.get(mnemonic.lower())
        if insn is None:
            raise RuntimeError('No {} instruction in instructions file.'
                               .format(mnemonic.upper()))

        if not self._check_insn_shape(insn, mnemonic.lower()):
            raise RuntimeError('Fetched instruction {} from instructions file '
                               'is not the shape expected by the generator.'
                               .format(mnemonic.upper()))

        return insn
```

**Review by rswarbrick — COMMENTED** [review]
This looks like it will be much cleaner, thanks!

I'm not sure about the giant dictionary lookup though: what do you think of the rejig in the comment?

File: hw/ip/otbn/dv/rig/rig/snippet_gen.py (hunk 1 above)
Comment (ctopal) [reply hw/ip/otbn/dv/rig/rig/snippet_gen.py]: Makes perfect sense honestly. Also helps with `bne` and `beq` being in the same shape! Thanks for the suggestion.

**Review by rswarbrick — APPROVED** [review]
Looks great, thanks!

**Pull request merged**